core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableBatchNestedLoopJoin.java

```diff
@@ -0,0 +1,252 @@
+import java.util.List;
+import java.util.Set;
+
+/** Implementation of batch nested loop join in
+ * {@link org.apache.calcite.adapter.enumerable.EnumerableConvention enumerable calling convention}. */
+public class EnumerableBatchNestedLoopJoin extends Join implements EnumerableRel {
+
+  private final ImmutableBitSet requiredColumns;
+  protected EnumerableBatchNestedLoopJoin(
+      RelOptCluster cluster,
+      RelTraitSet traits,
+      RelNode left,
+      RelNode right,
+      RexNode condition,
+      Set<CorrelationId> variablesSet,
+      ImmutableBitSet requiredColumns,
+      JoinRelType joinType) {
+    super(cluster, traits, left, right, condition, variablesSet, joinType);
+    this.requiredColumns = requiredColumns;
+  }
+
+  public static EnumerableBatchNestedLoopJoin create(
+      RelNode left,
+      RelNode right,
+      RexNode condition,
+      ImmutableBitSet requiredColumns,
+      Set<CorrelationId> variablesSet,
+      JoinRelType joinType) {
+    final RelOptCluster cluster = left.getCluster();
+    final RelMetadataQuery mq = cluster.getMetadataQuery();
+    final RelTraitSet traitSet =
+        cluster.traitSetOf(EnumerableConvention.INSTANCE)
+            .replaceIfs(RelCollationTraitDef.INSTANCE,
+                () -> RelMdCollation.enumerableBatchNestedLoopJoin(mq, left, right, joinType));
+    return new EnumerableBatchNestedLoopJoin(
+        cluster,
+        traitSet,
+        left,
+        right,
+        condition,
+        variablesSet,
+        requiredColumns,
+        joinType);
+  }
+
+  @Override public EnumerableBatchNestedLoopJoin copy(RelTraitSet traitSet,
+      RexNode condition, RelNode left, RelNode right, JoinRelType joinType,
+      boolean semiJoinDone) {
+    return new EnumerableBatchNestedLoopJoin(getCluster(), traitSet,
+        left, right, condition, variablesSet, requiredColumns, joinType);
+  }
+
+  @Override public RelOptCost computeSelfCost(
+      final RelOptPlanner planner,
+      final RelMetadataQuery mq) {
+    double rowCount = mq.getRowCount(this);
+
+    final double rightRowCount = right.estimateRowCount(mq);
+    final double leftRowCount = left.estimateRowCount(mq);
+    if (Double.isInfinite(leftRowCount) || Double.isInfinite(rightRowCount)) {
+      return planner.getCostFactory().makeInfiniteCost();
+    }
+
+    Double restartCount = mq.getRowCount(getLeft()) / variablesSet.size();
+
+    RelOptCost rightCost = planner.getCost(getRight(), mq);
+    RelOptCost rescanCost =
+        rightCost.multiplyBy(Math.max(1.0, restartCount - 1));
+
+    // TODO Add cost of last loop (the one that looks for the match)
+    return planner.getCostFactory().makeCost(
+        rowCount + leftRowCount, 0, 0).plus(rescanCost);
+  }
+
+  @Override public RelWriter explainTerms(RelWriter pw) {
+    super.explainTerms(pw);
+    return pw.item("batchSize", variablesSet.size());
+  }
+
+  public Result implement(EnumerableRelImplementor implementor, Prefer pref) {
```

Comment:
`@Override` annotation is missing.
Also, in this method (and in other methods of the class) there is a mix between local variables defined as `final`, and local variables not defined as `final` but which are effectively final, I think this should be homogenized.